chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -785,52 +842,105 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+	batchRequest := make(
+		map[uint32]rpcclient.FutureGetBlockHashResult, endHeight-startHeight,
+	)
+
 	for height := endHeight; height >= startHeight && height > 0; height-- {
+		sendReq, err := b.chainConn.GetBlockHashAsync(int64(height))
+		if err != nil {
+			return nil, fmt.Errorf("unable to receive the block "+
+				"promise %v", err)
+		}
+		batchRequest[height] = sendReq
+	}
+
+	// Sending the bulk request using updated batchClient.
+	b.chainConn.SendAsyncQueue()
+
+	blockHashAddresses := make([]*chainhash.Hash, 0, len(batchRequest))
+
+	for height, response := range batchRequest {
 		// Ensure we haven't been requested to shut down before
-		// processing the next height.
+		// processing the next height response.
 		select {
 		case <-b.quit:
 			return nil, chainntnfs.ErrChainNotifierShuttingDown
 		default:
 		}
 
 		// First, we'll fetch the block for the current height.
-		blockHash, err := b.chainConn.GetBlockHash(int64(height))
+		blockHash, err := response.Receive()
 		if err != nil {
 			return nil, fmt.Errorf("unable to retrieve hash for "+
 				"block with height %d: %v", height, err)
 		}
-		block, err := b.GetBlock(blockHash)
+
+		blockHashAddresses = append(blockHashAddresses, blockHash)
+	}
+
+	// Now we fetch blocks in interval of 15 to avoid out of memory errors
+	// in case of fetching 1000 blocks from GetBlocksBatch().
+	intervals := len(blockHashAddresses) / 15
+
+	for i := 0; i <= intervals; i++ {
+		// Ensure we haven't been requested to shut down before
+		// processing next set of blocks.
+		select {
+		case <-b.quit:
+			return nil, chainntnfs.ErrChainNotifierShuttingDown
+		default:
+		}
+		var (
+			blocks []*wire.MsgBlock
+			err    error
+		)
+		strt, end := i*15, i*15+15
+		if i == intervals {
+			if len(blockHashAddresses)%15 == 0 {
+				break
+			}
+			blocks, err = b.chainConn.GetBlocksBatch(
+				blockHashAddresses[strt:],
+			)
+		} else {
+			blocks, err = b.chainConn.GetBlocksBatch(
+				blockHashAddresses[strt:end],
+			)
+		}
 		if err != nil {
-			return nil, fmt.Errorf("unable to retrieve block "+
-				"with hash %v: %v", blockHash, err)
+			return nil, err
 		}
 
-		// Then, we'll manually go over every input in every transaction
-		// in it and determine whether it spends the request in
-		// question. If we find one, we'll dispatch the spend details.
-		for _, tx := range block.Transactions {
-			matches, inputIdx, err := spendRequest.MatchesTx(tx)
-			if err != nil {
-				return nil, err
-			}
-			if !matches {
-				continue
-			}
+		height := int(endHeight) - strt
+		for j := range blocks {
+			// Now we'll manually go over every input in every
+			// transaction in it and determine whether it spends the
+			// request in question. If we find one, we'll dispatch
+			// the spend details.
+			for _, tx := range blocks[j].Transactions {
+				matches, inputIdx, err := spendRequest.MatchesTx(tx)
+				if err != nil {
+					return nil, err
+				}
+				if !matches {
+					height--
```

Comment:
Again, this is incorrect, and perhaps better to return `if matches`.

Reply:
Sure